src/commands.rs

```diff
@@ -0,0 +1,38 @@
+use bevy::{
+    ecs::system::{Command, EntityCommands},
+    prelude::*,
+};
+
+/// Extension trait for [`EntityCommands`] that allows inserting a dynamic component.
+pub trait DynamicEntityCommandsExt {
+    /// Insert a component onto the entity that is not known at compile time.
+    fn insert_dynamic(
+        &mut self,
+        reflect_component: ReflectComponent,
+        component_data: Box<dyn Reflect>,
+    ) -> &mut Self;
+}
+
+struct InsertDynamicCommand(Entity, ReflectComponent, Box<dyn Reflect>);
+impl Command for InsertDynamicCommand {
+    fn write(self, world: &mut World) {
+        self.1.insert(world, self.0, self.2.as_reflect());
+    }
+}
+
+impl<'w, 's, 'a> DynamicEntityCommandsExt for EntityCommands<'w, 's, 'a> {
+    fn insert_dynamic(
+        &mut self,
+        reflect_component: ReflectComponent,
+        component_data: Box<dyn Reflect>,
+    ) -> &mut Self {
+        let entity = self.id();
+        self.commands().add(InsertDynamicCommand(
+            entity,
+            reflect_component,
+            component_data,
+        ));
+
+        self
+    }
+}
```

Comment:
I think I may have missed how/where the requirement for dynamic components was coming in, or misattributed to scripting when its required for the states in general?

Reply:
Yes the dynamic components are actually used in states, regardless and totally unrelated to scripting.

Though it does make scripted attacks _way_ easier to implement later, it is still absolutely useful and utilized in the current state system completely irrespective of any scripting effort.

It's actually kind of a good sign that you missed it, at least as far as readability state transition logic is concerned. I'm glad the new concept didn't make the state transition systems' implementations hard to understand.

I'll walk through it and why it helps.

---

When we want to express the intent to transition to a new state we do something like this:

```rust
if action_state.just_pressed(PlayerAction::Attack) {
    transition_intents.push_back(StateTransition::new(
        Attacking::default(),
        Attacking::PRIORITY,
        false,
    ));
}
```

We express the desired to add an `Attacking` component to the fighter, and we push it into the `VecDeque`.

Later in the same function we do this:

```rust
transition_intents.push_back(StateTransition::new(
    Throwing,
    Throwing::PRIORITY,
    true,
));
```

Here we express the desire to add a _different_ component to the fighter, but we are pushing the intent to the _same_ `VecDeque`!

This means we have to "erase" the component type so that we can store multiple different component types in the same `VecDeque`. This is conceptually similar to needing to do a `Vec<Box<dyn Any>>` if you want to store multiple different types in the same vector.

This type erasure is manged by the `StateTransition::new` function. It boxes the value and erases it's type while also querying the Bevy type registry to get the info needed to add the component to an entity later. And all that gets stored in the `StateTransition`, along with the `priority` and `is_additive` options.

Finally, when we are looping through state transition intents and deciding which one to transition to we do this:

```rust
if intent.priority > current_state_priority {
    // Apply the state
    intent.apply::<CurrentState>(entity, commands);
}
```

We are saying that we want to take the component stored in the `StateTransition` `intent` and add it to the fighter, even though we don't know what component that is, because it's been type erased.

After my latest commit, to get rid of the confusing commands stuff, `StateTransition::apply` looks like this:

```rust
pub fn apply<CurrentState: Component>(self, entity: Entity, commands: &mut Commands) -> bool {
    if !self.is_additive {
        commands.entity(entity).remove::<CurrentState>();
    }

    commands.add(move |world: &mut World| {
        // Insert the component stored in this state transition onto the entity
        self.reflect_component
            .insert(world, entity, self.data.as_reflect());
    });

    self.is_additive
}
```

The `reflect_component` has the type information needed to insert a new instance of the type-erased component onto the entity with the `insert` function. And we must pass it the `world`, the `entity` to add the component to, and our `Box<dyn Reflect>` which contains our actual component data.

---

That completes the workflow and allows us to store a bunch of transitions to any kind of new state in a single queue. And using components instead of an enum allows us to use the awesome Bevy component filtering syntax in our queries.

All that has "nothing" to do with scripting. But it makes it does make it trivial for a script to decide to emit a state transition intent for a new, unknown state, that is handled by the script. As long as the priority is higher than the existing state, the state will transition in the Rust system, and at that point the script state handler takes over and can control the player however it wants at that point. Voilà! You have a new attack/status effect/whatever without modifying core.

And this pattern, again, is helpful even if we only ever wrote Rust.

Reply:
Yep seems good. Very valuable clear explanation. I wonder if it would be good to retain this doc somewhere, seems useful in case someone wants to understand whats going on under the hood.

I caught the way we are using VecDeque but was missing the part about what was stored inside being type erased. Although I have a basic understanding of Boxing/type erasure I've never actually used it, and totally missed that we were storing an actual component in the state transition.

Reply:
I think that about clears up the last of my questions. Probably good to merge whenever you are ready.

Reply:
> I wonder if it would be good to retain this doc somewhere, seems useful in case someone wants to understand whats going on under the hood.

Great idea. I'll move it to the Wiki as a sub-page under Architecture probably, and I'll also write up an explanation of the high-level Architecture after merging.